### PR TITLE
fileset: enable glob matching by default

### LINKED ISSUE #6971: FR: Is the `glob:` prefix necessary for filesets?
Coming recently from `git`, I always instinctively try to use "bare" globs, e.g.

```
jj restore --from=main '**/settings.py'
```

Only to see that `Nothing changed.`, and then remember that `jj` requires the `glob:` prefix.

Is this requirement a technical limitation of the fileset implementation, or would it be possible to support bare glob syntax? This is something I do every day, and having to use the `glob:` prefix every time is a bit cumbersome.

I can learn to deal with it until it becomes muscle memory, but I thought I'd raise a feature request first to understand if the bare glob syntax could be supported.

### PR DESCRIPTION
**Pull request by yuja**
Closes #6971

# Checklist

If applicable:

- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [x] I have added/updated tests to cover my changes

### CODE REVIEW

File: lib/src/matchers.rs [hunk 1]
```diff
@@ -45,6 +45,12 @@ pub enum Visit {
 }
 
 impl Visit {
+    /// Entries in the directory will match, but the details are unknown.
```
Comment (martinvonz) [inline lib/src/matchers.rs]: The entries need to be visited but they don't necessarily match, right? Ah, maybe you meant "*Some* entries ..." but I read it as "*All* entries". I think expressing it as "All entries in the directory need to be visited but they may not all match" or similar.

(Also, in your sentence, I think the "will" perhaps should be a "may" to allow for somewhat lazy/sloppy implementation that doesn't restrict visits as much as it could.)

File: lib/src/matchers.rs [hunk 2]
```diff
@@ -774,10 +823,13 @@ mod tests {
 
     #[test]
     fn test_file_globs_matcher_wildcard_any() {
-        // "*" could match the root path, but it doesn't matter since the root
-        // isn't a valid file path.
+        // It's not obvious whether "*" should match the root directory path.
+        // Since "<dir>/*" shouldn't match "<dir>" itself, we can consider that
+        // "*" has an implicit "<root>/" prefix, and therefore it makes sense
+        // that "*" doesn't match the root. OTOH, if we compare paths as literal
+        // strings, "*" matches "". The current implementation is the former.
```
Comment (martinvonz) [inline lib/src/matchers.rs:831]: SGTM. That's what I would expect.

File: lib/src/matchers.rs [hunk 3]
```diff
@@ -800,6 +852,120 @@ mod tests {
...
+        let m = new_prefix_globs_matcher(&[(RepoPath::root(), glob("*.rs"))]);
+        assert!(!m.matches(repo_path("foo")));
+        assert!(m.matches(repo_path("foo.rs")));
+        assert!(m.matches(repo_path("foo\n.rs"))); // "*" matches newline
+        assert!(!m.matches(repo_path("foo.rss")));
+        assert!(m.matches(repo_path("foo.rs/bar")));
+        assert!(!m.matches(repo_path("foo/bar.rs")));
+        assert_eq!(m.visit(RepoPath::root()), Visit::SOME);
+        assert_eq!(m.visit(repo_path("foo.rs")), Visit::AllRecursively);
+        assert_eq!(m.visit(repo_path("foo.rs/bar")), Visit::AllRecursively);
```
Comment (martinvonz) [inline lib/src/matchers.rs:867]: Also test a negative match?

File: cli/src/commands/file/show.rs
Comment (martinvonz) [inline cli/src/commands/file/show.rs:1]: > It's odd that literal patterns do prefix matching by default whereas globs match
exactly. We can instead rename the existing "glob:" to "file-glob:" and add
prefix "glob:", but I'm not sure which is better.

I think what you have sounds good. If someone explicitly uses `glob:foo`, they might be surprised if it matches recursively.

File: CHANGELOG.md [hunk 4]
```diff
@@ -10,6 +10,8 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### Breaking changes
 
+* In filesets or path patterns, glob matching is enabled by default.
```
Comment (martinvonz) [inline CHANGELOG.md]: nit: Might be good to say here that users can use the `cwd:` to get the old behavior (even though I think it's rarely going to be useful).

File: lib/src/matchers.rs (hunk 1 above)
Comment (yuja) [reply lib/src/matchers.rs]: Updated to
```
    /// All entries in the directory need to be visited, but they are not
    /// guaranteed to match.
```

File: cli/src/commands/file/show.rs
Comment (yuja) [reply cli/src/commands/file/show.rs:1]: Thanks. Removed TODO about `glob:` naming.

File: lib/src/matchers.rs (hunk 3 above)
Comment (yuja) [reply lib/src/matchers.rs:867]: Added `foo.rss` and `foo.rss/bar`, but they are `Visit::SOME` because our matcher isn't clever enough to determine that `*.rs` never matches child entries.
